### PR TITLE
[not essential@beta0] exported CYLC_TEMPLATE_VARS

### LINKED ISSUE #3988: templatevars: add dictionary of all templatevars
Rose provides a dictionary containing all template variables called `ROSE_SUITE_VARIABLES`.

For example this config:

```
[jinja2:suite.rc]
ANSWER=42
```

Will result in this `ROSE_SUITE_VARIABLES` dict:

```python
ROSE_SUITE_VARIABLES = {
    'ANSWER': 42
}
```

This functionality is specific to Rose so if you override a variable on the Cylc command line it will not effect the Rose output:

```bash
# this will set ANSWER to 43, however the value in ROSE_SUITE_VARIABLES will still be 42
cylc validate -s ANSWER=43
```

This is a bit messy, however, there is a simple solution, just add this functionality into Cylc. Export a dictionary called `CYLC_TEMPLATE_VARS` or something of that ilk.

Tagged against 8.x as this is not essential for 8.0.0, however, it should be pretty quick to implement so may be done before then.

### PR DESCRIPTION
**Pull request by wxtim**
These changes close #3988 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.

- [x] (master branch) I have opened a documentation issue. I believe that this needs to be added as part of a general update to documentation of the template variables handling. It doesn't sit nicely within the existing structure.
- [x] No dependency changes.

### CODE REVIEW

**Comment by hjoliver** [comment]
@wxtim so for (e.g.) `cylc validate -s "FOO='foo'" flow.cylc` we should be able to do this:
```
# flow.cylc
# FOO is {{CYLC_TEMPLATE_VARS["FOO"]}}
```
as well as this:
```
# flow.cylc
# FOO is {{FOO}}
```
?

**Comment by oliver-sanders** [comment]
Yes, although we wouldn't advise people to write `CYLC_TEMPLATE_VARS["FOO"]` rather than just `FOO` of course.

This dictionary is a convenient way to see all of the template variables that have been set, iterate/search them which has some uses. This is analogous to / replaces the `ROSE_SUITE_VARS` dictionary that no longer makes much sense.

**Review by MetRonnie — COMMENTED** [review]
(Oops, didn't press submit review)

File: tests/functional/validate/75-CYLC_TEMPLATE_VARS.t [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+[runtime]
+  [[root]]
+    script = true
+  [[another]]
+  [[one_thing]]
+__HEREDOC__
+
+run_ok "${TEST_NAME_BASE}-validate" cylc validate .
+grep_ok "CYLC_TEMPLATE_VARS" "${TEST_NAME_BASE}-validate.stderr"
+grep_ok "CYLC_VERSION" "${TEST_NAME_BASE}-validate.stderr"
```
[the remark this answers is not in the document]
Comment (wxtim) [reply tests/functional/validate/75-CYLC_TEMPLATE_VARS.t:42]: > I'm in favour of removing every unnecessary thing from tests, as they can distract or confuse down the line

If it's not trivial I'm  in favour of using the `[meta]` section of the test suite to explain what it does.

[the remark this answers is not in the document]
Comment (wxtim) [reply tests/functional/validate/75-CYLC_TEMPLATE_VARS.t]: I think I've both streamlined and improved the test..

**Review by MetRonnie — APPROVED** [review]

**Review by hjoliver — APPROVED** [review]

**Pull request merged**